consensus/hotstuff/cruisectl/block_time_controller.go

```diff
@@ -450,17 +430,25 @@ func (ctl *BlockTimeController) measureViewDuration(tb TimedBlock) error {
 
 // processEpochExtended processes the EpochExtended protocol data.
 // Whenever we encounter an epoch extension, we:
-//   - set epochFallbackTriggered to true if this is the first extension encountered.
 //   - update the curr epoch final view and target end time with the extension data.
 //
 // No errors are expected during normal operation.
-func (ctl *BlockTimeController) processEpochExtended(_ uint64, _ *flow.Header, extension flow.EpochExtension) error {
-	if !ctl.epochFallbackTriggered {
-		// set epochFallbackTriggered to true when we encounter the first extension
-		ctl.epochFallbackTriggered = true
+func (ctl *BlockTimeController) processEpochExtended(header *flow.Header, extension flow.EpochExtension) error {
+	// sanity check: ensure the final view of the current epoch monotonically increases
+	if extension.FinalView < ctl.epochInfo.curEpochFinalView {
+		return fmt.Errorf("final view of epoch must be monotonically increases, but is decreasing from %d to %d", ctl.epochInfo.curEpochFinalView, extension.FinalView)
+	}
+
+	if extension.FinalView == ctl.epochInfo.curEpochFinalView {
+		return nil
 	}
+
 	ctl.curEpochFinalView = extension.FinalView
-	ctl.curEpochTargetEndTime = extension.TargetEndTime
+	targetEndTime, err := ctl.state.AtBlockID(header.ID()).Epochs().Current().TargetEndTime()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve target end time at block id %s", header.ID())
+	}
+	ctl.curEpochTargetEndTime = targetEndTime
```

Comment:
```suggestion
	ctl.curEpochTargetEndTime = targetEndTime
	ctl.curEpochTargetDuration = targetDuration
```
We also need to update target duration here. We can get this value from the protocol state as well.
```go
ctl.state.AtHeight(header.Height).Epochs().Current().TargetDuration()
```

Let's also update [the test](https://github.com/onflow/flow-go/blob/a221aee99638fd22c68156f45e2ceff6e9f734ed/consensus/hotstuff/cruisectl/block_time_controller_test.go#L253) to check that this value is updated

Reply:
https://github.com/onflow/flow-go/pull/6102/commits/3c7d837f5cf62d401c7e9c894a7608e6e7a1f7ad